### PR TITLE
OBJ-200 real download client method

### PR DESCRIPTION
**Pull request by mattch1**
Download call to the file transfer api based on extensions but refactored. 
Included a soluton to check the metadata for the av status - as I was doing this I noticed a change to the criteria now excluding this - so I have committed what I have done for people to look at in case we might still want to check first - if so I can add link it to the code and add unit tests and if not I can just remove it.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java [hunk 1]
```diff
@@ -27,7 +29,12 @@
     private static final String UPLOAD = "upload";
     private static final String CONTENT_DISPOSITION_VALUE = "form-data; name=%s; filename=%s";
     private static final String NULL_RESPONSE_MESSAGE = "null response from file transfer api url";
+    private static final String GET_URI = "%s/%s";
```
Comment (ICoxCH) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:32]: this and delete uri are the same value, would it be possible to merge the constants?

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:32]: We could but it is more readable this way and if one were to change.. so I've kep them separate.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java [hunk 2]
```diff
@@ -167,15 +175,84 @@ public FileTransferApiClientResponse delete(String requestId, String fileId) {
...
+    private ClientHttpResponse getClientHttpResponse(HttpServletResponse httpServletResponse, ClientHttpResponse clientHttpResponse) throws IOException {
+        setResponseHeaders(httpServletResponse, clientHttpResponse);
 
-        InputStream inputStream = new ByteArrayInputStream("This is a test".getBytes());
+        InputStream inputStream = clientHttpResponse.getBody();
         IOUtils.copy(inputStream, httpServletResponse.getOutputStream());
+        return clientHttpResponse;
+    }
 
+    private FileTransferApiClientResponse getFileTransferApiClientResponse(ClientHttpResponse clientHttpResponse) throws IOException {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java]: Not sure, but do you need to declare the 'throws IOException' on this method. Try removing it and see if code compiles.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java [hunk 3]
```diff
@@ -167,15 +175,84 @@ public FileTransferApiClientResponse delete(String requestId, String fileId) {
...
+
+                //FileTransferResponseBuilder - the output from FileTransferOperation is the input into
+                // this FileTransferResponseBuilder
+                clientHttpResponse -> {
+                    return getFileTransferApiClientResponse(clientHttpResponse);
+                }
+        );
+    }
+
+    private boolean isFileClean(String requestId, String fileId) {
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:193]: Have you tried testing locally and including this call with both a clean and an infected file? Just interested...

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:193]: I would need an infected file in order to do this.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java (hunk 2 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java]: No, won't compile without them

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java (hunk 3 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:193]: 'Safe' infected files that contain a virus signature are available. Maybe speak to test team who might know what to use.

**Comment by ICoxCH** [comment]
This pr still needs increased unit test coverage, primarily on the new code in the file transfer api client

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java [hunk 4]
```diff
@@ -167,15 +174,76 @@ public FileTransferApiClientResponse delete(String requestId, String fileId) {
...
+
+    }
+
+    private void handleRequestCallback(ClientHttpRequest requestCallback) {
+        requestCallback.getHeaders().setAccept(Collections.singletonList(MediaType.APPLICATION_OCTET_STREAM));
+        requestCallback.getHeaders().add(HEADER_API_KEY, fileTransferApiKey);
+    }
 
-        InputStream inputStream = new ByteArrayInputStream("This is a test".getBytes());
+    private ClientHttpResponse getClientHttpResponse(HttpServletResponse httpServletResponse, ClientHttpResponse clientHttpResponse) throws IOException {
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:214]: this method looks like it is doing the copying, the getClientHttpResponse name doesn't indicate that any copying will take place. Also we are passing a ClientHttpResponse in to the getClientHttpResponse which is confusing.

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:214]: Perhaps "copyClientHttpDataToServletResponse" would be better

**Comment by mattch1** [comment]
The reason why the test coverage is low is because I'm not sure whether we want to use what I have written for av check or not - see my initial comment asking for a decision on this.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java [hunk 5]
```diff
@@ -167,15 +174,76 @@ public FileTransferApiClientResponse delete(String requestId, String fileId) {
...
+                        clientHttpResponse -> getClientHttpResponse(httpServletResponse, clientHttpResponse)
+                        ),
+                clientHttpResponse -> getFileTransferApiClientResponse(requestId, clientHttpResponse)
+        );
+    }
+
+    private boolean isFileClean(String requestId, String fileId) {
+        String getUrl = String.format(FILE_TRANSFER_URI_TEMPLATE, fileTransferApiURL, fileId);
+
+        FileTransferApiClientResponse fileDataResponse = makeApiCall(
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:196]: can you put the av checking stuff onto another branch so we don't lose it?

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java [hunk 6]
```diff
@@ -167,15 +174,76 @@ public FileTransferApiClientResponse delete(String requestId, String fileId) {
...
+                requestId,
+                () -> {
+                    HttpEntity<Void> request = new HttpEntity<>(createApiKeyHeader());
+                    return restTemplate.exchange(getUrl, HttpMethod.DELETE, request, FileTransferApiResponse.class);
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java:200]: you've got HttpMethod.DELETE, we don't want to delete the file we just want to get the file details.

**Comment by markpit** [comment]
> The reason why the test coverage is low is because I'm not sure whether we want to use what I have written for av check or not - see my initial comment asking for a decision on this.

We're not going to implement the av check - it was in our slack channel and i've updated the task to reflect this

**Comment by markpit** [comment]
didn't mean to close, pressed wrong button, should be open again...

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java (hunk 6 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java]: Removed method - did set to get must have ctrl-z reverse this with another change.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java (hunk 5 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java]: Will do once this is done and merged in.

File: src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java (hunk 4 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/api/strikeoffobjections/file/FileTransferApiClient.java]: Changed name.

**Pull request merged**